doc/adr/0009-container-image-tags.md

```diff
@@ -0,0 +1,52 @@
+# 9. Convention for tagging of container images
+
+Date: 2020-03-13
+
+## Status
+
+Proposed
+
+## Context
+
+Libero's products are assembled from product specific, general libero and external container images.
+
+Integration testing and reproducible deploys require uniquely identifiable image tags.
+
+As a starting point for users each product has an umbrella repo. This is not only used to collect product issues and documents. I also contains e.g. docker-compose files, helm charts that combine various components to a useable whole.
+
+
+## Decision
+
+### Tagging 
+
+- libero container images shall always be tagged with {branch-full_git_hash}
```

Comment:
why do want to include the branch in the tag?

Reply:
In early Libero discussions it was proposed to add suffix such as `_unstable` for non-release tags. The reason was mostly because Docker Hub would get quite crowded and it would be difficult to differentiate the two. I am not sure whether that was actually used for Libero projects (@giorgiosironi will know). (I used that for ScienceBeam and other eLife projects I was involved in)

Reply:
This would let downstream consumers of the image e.g. integration testing, deployments track images from a specific branch. This would leave it open to product and component teams to choose and adapt their branching and tagging strategy.

For instance, as of now there doesn't appear to be a high frequency of semver tagging going on as we are in the early stages for various components. As such it would make sense from my perspective to track master-branch images for integration testing.

Another example might be a breaking change introducing major version development of a component that needs to run parallel to ongoing bug-fixes etc. We could then have two integration test branches or multiple staging environments.

Reply:
> This would let downstream consumers of the image e.g. integration testing, deployments track images from a specific branch. This would leave it open to product and component teams to choose and adapt their branching and tagging strategy.
> 
> For instance, as of now there doesn't appear to be a high frequency of semver tagging going on as we are in the early stages for various components. As such it would make sense from my perspective to track master-branch images for integration testing.
> 
> Another example might be a breaking change introducing major version development of a component that needs to run parallel to ongoing bug-fixes etc. We could then have two integration test branches or multiple staging environments.

Just checking that all of that is response to the question regarding including the branch in the tag?

Reply:
> to add a image repo suffix such as _unstable for non-release tags

Just add a call with @erkannt to prepare this for the Libero sync meeting; the intention of this branch name use is to address the same concern.

In the immediate future, I would expect only `master-*` tags to appear but it leaves open both:
- the possibility to use a `develop` branch 
- to publish `my-feature-branch-*` tags for testing inside a Kubernetes cluster (or anything else that needs images pullable)

As much as `_unstable` image suffixes help in the Docker Hub UI, the idiomatic way for Docker seems to be to have a single image name and way too many tags rather than multiple image names.

Reply:
Ideally speaking we wouldn't rebuild an image if one already exists for that git-hash. But this will be branching, merging etc. specific. It shouldn't hurt if we rebuild from the same git-hash and if it does I am confident that solutions are possible.

Regarding the use of tags: the motivation of this ADR is to enable reproducible integration test and more importantly deploys.

- git tags trigger tagging of container images
- image digests are used by continuous deploy tools as unique identifier of images

The latter becomes relevant if you e.g. want pin the deployed image to a major release number but automatically upgrade minor and bugfixes. In this case your image tag would be constant (e.g. nginx:17) but its digest would change making PRs understandable and leading to reproducibility.